libs/constrained/src/Constrained/GenT.hs

```diff
@@ -0,0 +1,171 @@
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE DefaultSignatures #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE DeriveFunctor #-}
+{-# LANGUAGE ViewPatterns #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE EmptyCase #-}
+{-# LANGUAGE ImportQualifiedPost #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE DerivingVia #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TupleSections #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE QuantifiedConstraints #-}
+module Constrained.GenT where
+
+import GHC.Stack
+import Control.Monad
+import Test.QuickCheck hiding (Fun, Args)
+import Test.QuickCheck.Gen
+import Test.QuickCheck.Random
+
+data GenMode = Loose
+             | Strict
+             deriving (Ord, Eq, Show)
+
+newtype GenT m a = GenT { runGenT :: GenMode -> Gen (m a) }
```

Comment:
Why don't we use an existing implementation `GenT`: https://hackage.haskell.org/package/quickcheck-transformer-0.3.1.2/docs/Test-QuickCheck-GenT.html#t:GenT

This could then be: `GenT (ReaderT GenMode m) a`. It would let you avoid redefining bunch of functions and instances in this module.


We could even define a stack along the lines of:
```
GenT (ReaderT GenMode (WriterT [String] (ExceptT [String] m))) a
```

`WriterT` would be used for failures, `EsceptT` for fatal failures

Reply:
I've looked some more at this and I don't think it's worth it, unfortunately.

The problem is that we have this strictness parameter (`GenMode`) that matters for the semantics of all the functions you would normally get "out of the box" with `GenT`. Consequently, what you would end up having to do is to:
- Implement our `GenT` as a newtype wrapper around `ReaderT GenMode (GenT m)`
- Not expose the newtype wrapper
- Not implement the `MonadGen` interface
- Get rid of our instances of `Applicative` and `Monad`
- Possibly simplify `pureGen`

As far as I can tell only the latter two are wins here, but at the cost of making the code more complicated / confusing.

Reply:
It is worth it. You just don't have the full picture of ledger in mind. 

By using `quickcheck-transformer` we would be able to write custom generators using the `MonadGen` type class in other test suites in the ledger codebase and be able reuse them here. 

But, that's fine, we can solve this problem later, feel free to keep current definition as it is.

Reply:
Read what I wrote again, we can't provide an instance for `MonadGen` for this type because that would break the intended behaviour. Therefore, it does not give us any reusability.

Reply:
> that would break the intended behaviour.

Could you be a little bit more specific?

Reply:
> The problem is that we have this strictness parameter (GenMode) that matters for the semantics of all the functions you would normally get "out of the box" with GenT.

C.f. suchThatT for example.

Reply:
For instance, `listOfT` switches the `GenMode` to `Loose` when generating the elements (it's ok to fail occasionally when generating a list, you just get fewer elements). Using `listOf` from `MonadGen` would not have this behaviour. Most of the interesting combinators interact with the `GenMode` in some way.